### ISSUE TITLE
In GAIL and AIRL networks, use logits_expert_is_high instead of logits_gen_is_high

### ISSUE THREAD

**Issue by shwang** [HumanCompatibleAI/imitation#275]
For consistency with GAIL/AIRL paper implementation and equations.

We currently use discriminator logits that are high if they predict a particular sample is from the generator, rather than the expert (reason: I chose the wrong arbitrary sign when I first made this repo). By reversing the meaning of our logits we will better align with "the literature" and won't have to [run into confusion when editing the code](https://github.com/HumanCompatibleAI/imitation/pull/198#discussion_r444539134).

**Comment by Rocamonde** [comment]
Working on this, finished for AIRL. Not sure what is going on for the GAIL case: https://github.com/HumanCompatibleAI/imitation/blob/2a383ebfeef1753a9538435cb35e7b1f7f0ddb80/src/imitation/algorithms/adversarial/gail.py#L97-L99
It's only returning the $r_\theta(s,a)$ value as the logits, is this correct / how it works in the paper? It does not seem consistent with my initial exploration of the literature, but still making my way through the maths, so could be totally wrong. 

Also, does this not already assume that "expert is high" rather than "gen is high" as was allegedly the convention? But the optimization logic which I just re-wrote on https://github.com/HumanCompatibleAI/imitation/commit/5191af2c99672e3a7673abbad82f440bff8cb284 actually followed the opposite convention - so how did this actually manage to work in training?

**Comment by AdamGleave** [comment]
Thanks for making a start on this.

The GAIL thing is confusing, and does seem slightly off, although in a way that I can see might still work. In any case, we should definitely clear this up.

My understanding is that `self._reward_net`, trained under the current generator-is-high convention, corresponds to `\log(D(s,a))` in the [GAIL paper](https://arxiv.org/pdf/1606.03476.pdf)  (see eq. 16). I think this part is OK, and the actual discriminator training should work, although it's the opposite of what one might expect from the name "reward net" (higher reward is *worse*...).

Now `self.reward_train` is `self._processed_reward`, which is `LogSigmoidRewardNet(self._reward_net)`. Despite the name, actually `LogSigmoidRewardNet` actually returns the [*negative* log sigmoid](https://github.com/HumanCompatibleAI/imitation/blob/2a383ebfeef1753a9538435cb35e7b1f7f0ddb80/src/imitation/algorithms/adversarial/gail.py#L36). So `LogsigmoidRewardNet(x)` returns `-log(1/(1+(exp(-x))=\log(1 + exp(-x))`.

Now, substituting `x = \log(D(s,a))`, we get `\log(1 + 1/(D(s,a))`. The GAIL paper says we should use "cost function" `\log(D(s,a))` (step 5 of Algorithm 1), which should equate to reward function `\log(1/D(s,a))$. 

Now, that's not the same as `\log(1 + 1/D(s,a))`. But it's close, and they're at least monotonically related! So I guess it's close enough to work OK empirically? But we should probably match the paper unless we have a reason to the contrary. Tagging @shwang who developed this code originally in case there's a consideration I'm missing.

I'm not 100% on this, I changed my interpretation a couple of times looking into this, and don't have time to double-check everything tonight. If you also end up confused then let me know and I can take another look tomorrow!
